### PR TITLE
Relax `DROP PART` guarantees and turn on checks in ReplicationQueue.

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix an extremely rare bug which can lead to intersecting parts after `DROP PART` or background deletion of an empty part.

### CODE REVIEW

**Comment by alesapin** [comment]
@Mergifyio update

**Comment by alesapin** [comment]
@Mergifyio update

**Comment by alesapin** [comment]
No, `DROP PART` still broken...

**Comment by alesapin** [comment]
@mergifyio update

File: src/Storages/MergeTree/DropPartsRanges.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    MergeTreePartInfo entry_info = MergeTreePartInfo::fromPartName(*entry->getDropRange(format_version), format_version);
+    drop_ranges.emplace(entry->znode_name, entry_info);
+}
+
+void DropPartsRanges::removeDropRange(const ReplicatedMergeTreeLogEntryPtr & entry)
+{
+    if (entry->type != ReplicatedMergeTreeLogEntry::DROP_RANGE)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to remove entry of type {} from drop ranges, expected DROP_RANGE", entry->typeToString());
+
+    drop_ranges.erase(entry->znode_name);
```
Comment (tavplubix) [inline src/Storages/MergeTree/DropPartsRanges.cpp:49]: Maybe add assertion that checks that `drop_ranges` had actually contained `entry->znode_name`?

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MergeTree/DropPartsRanges.cpp (hunk 1 above)
Comment (alesapin) [reply src/Storages/MergeTree/DropPartsRanges.cpp:49]: ok.

**Comment by alesapin** [comment]
no related failures.

**Pull request merged**